include/pcp/filter/density_filter.hpp

```diff
@@ -0,0 +1,115 @@
+﻿#ifndef PCP_FILTER_DENSITY_FILTER_HPP
+#define PCP_FILTER_DENSITY_FILTER_HPP
+
+/**
+ * @file
+ * @ingroup filter
+ */
+
+#include <algorithm>
+#include <iterator>
+#include <pcp/algorithm/average_distance_to_neighbors.hpp>
+#include <vector>
+
+namespace pcp {
+namespace density_filter {
+
+/**
+ * @brief parameters for the filter, density threshold is
+ * @tparam ScalarType Scalar type for the parameter variables (float,double,etc)
+ */
+template <class ScalarType>
+struct construction_params_t
+{
+    using scalar_type              = ScalarType;
+    scalar_type density_threshold_ = static_cast<float>(1.0f);
+    scalar_type radius_multiplier_ = static_cast<float>(1.0f);
+};
+} // namespace density_filter
+
+/**
+ * @brief Iterates through the entire input once, calculates the density of the points for every
+ * point and removes points that do not pass the density threshold passed in the parameters.
+ * 
+ * @tparam Element Element Type
+ * @tparam ParametersType type of the parameters (double,float)
+ * @tparam PointMap Type satisfying PointMap concept
+ * @tparam KnnMap Type satisfying KnnMap concept
+ * @tparam RangeSearchMap Type satisfying Range concept
+ */
+template <class Element, class ParametersType, class PointMap, class KnnMap, class RangeSearchMap>
+class basic_density_filter_t
```

Comment:
Should not prefix with "basic_". We use the prefix "basic_" when there is an alias or typedef for the default template specialization of the basic type. For example, std::string is an alias for std::basic_string<char>. We used pcp::basic_linked_octree because there is a default alias for pcp::octree_t, for example.